Emo 6145

sujithvaddi opened the pull request:
## Github Issue #

<Create a new Github Issue and link here>

## What Are We Doing Here?
* Tweak the emo compaction process during the stash run.

## How to Test and Verify

1. Check out this PR
2. Run Command X, Click Button Y
3. Profit

## Risk

### Level 

`Low`, `Medium`, or `High`. Give an indication of what you think is the level of change introduced by this PR. `High` means a massive change to a core functionality. 
`Low` means a really minor change that shouldn't have any regression effect. 

### Required Testing

`Smoke`, `Regression`, or `Manual`. (All changes except documentation need smoke
testing at a minimum).

### Risk Summary

Add one or a few complete sentences about the possible risks or concerns for
this change.

## Code Review Checklist

- [ ] Tests are included. If not, make sure you leave us a line or two for the reason.

- [ ] Pulled down the PR and performed verification of at least being able to
build and run.

- [ ] Well documented, including updates to any necessary markdown files. When
we inevitably come back to this code it will only take hours to figure out, not
days.

- [ ] Consistent/Clear/Thoughtful? We are better with this code. We also aren't
a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.

- [ ] PR has a valid summary, and a good description.

sujithvaddi commented:
DO NOT MERGE YET - Its only for early feedback, it's not complete.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/system/SystemSource.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.Map;
+
+public interface SystemSource {
```
There's a missing implementation of this.  The two implementations provided either update the state in the local data center or call to a remote instance and update it's state.  What needs to happen is both.  For example, when a Stash starts in us-east-1 it needs to update the local state in us-east-1 AND make a call to eu-west-1 and update its state using the resource endpoint.  Unless both take place there is no guarantee that compaction would be restricted while Stash is running.

billkalter commented on web/src/test/java/com/bazaarvoice/emodb/web/scanner/ScanUploaderTest.java:
```diff
@@ -307,11 +309,15 @@ public S3ScanWriter answer(InvocationOnMock invocation)
         ScanStatusDAO scanStatusDAO = new InMemoryScanStatusDAO();
         // Create the instance and run the upload
-        ScanUploader scanUploader = new ScanUploader(dataTools, scanWorkflow, scanStatusDAO, stashStateListener);
+        ScanUploader scanUploader = new ScanUploader(dataTools, scanWorkflow, scanStatusDAO, stashStateListener, dataCenters, metricRegistry);
```
This unit test won't work because `ScanUploader` needs to have `DefaultSystemManager` injected.  Otherwise the default implementation always tries to open a connection to a remote host which does not exist.  With injection you can provide a mock and verify the correct calls are made.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/resources/system/SystemResource1.java:
```diff
@@ -0,0 +1,61 @@
+    @Path ("/stash-time/{id}")
+    // @RequiresPermissions ("system|update")
+    public SuccessResponse updateStashTime(@PathParam ("id") String id,
+                                           @QueryParam ("timestamp") LongParam timestampInMillis) {
```
It would be nice if this also included a list of placements.  For example, if a stash is running in us-east-1 there is no reason to halt compaction on eu-west-1-only placements such as `ugc_eu`.  Similarly, if stash in us-east-1 is only running for `app_global` there is no need to halt compaction on other global placements besides that one.

sujithvaddi commented:
- Now everything is named after compaction control instead of the generic ’system’ term.
- Added Permissions on the API and the required Authentication.
- API is more RESTful now.
- Transformed the timestamps to ISO8601.
- Changed the interface to also include placements, datacenter, remote along with the timestamp.
- Updated Guice Injections.   
- Added the missed deletion stash time call in Local Scan monitor.
- Now, stash running in one datacenter can update the stash times in all the data centers.
- Fixed ScanUploaderTests.

sujithvaddi commented:
@billkalter Please give me your comments when you have a chance.

billkalter commented on sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/AstyanaxDataWriterDAO.java:
```diff
@@ -277,7 +288,10 @@ public void compact(Table tbl, String key, UUID compactionKey, Compaction compac
 
-            deleteCompactedDeltas(rowKey, consistency, placement, keyspace, changesToDelete, historyList, tbl, key);
+            DateTime cutoffTimeForDeletes = CompactionControlUtils.getOldestStashStartTime(_defaultCompactionControlManager, _dataCenters);
+            if (cutoffTimeForDeletes == null || cutoffTimeForDeletes.getMillis() > System.currentTimeMillis()) {
```
I think this check makes more sense here in https://github.com/bazaarvoice/emodb/blob/master/sor/src/main/java/com/bazaarvoice/emodb/sor/core/DistributedCompactor.java#L52

billkalter commented on sor/src/main/java/com/bazaarvoice/emodb/sor/core/DefaultDataStore.java:
```diff
@@ -334,12 +339,16 @@ private Resolved resolve(Record record, ReadConsistency consistency) {
-        return expand(record, fullConsistencyTimeStamp, rawConsistencyTimeStamp, ignoreRecent, consistency);
+        // TODO: right now going with the old stash time just in the current data center - right thing is to get this from all the datacenters.
+        // (Need to address this after confirming if the current data centers approach is the best way to go.)
+        long stashRunTimeStamp = _compactionControlSource.getOldStashTime();
```
This is overly broad since it considers the oldest stash time for all placements.  For example, say there is only one Stash going on for placement `ugc_global:ugc` but the record being expanded is in `ugc_eu:ugc`.  There's no reason to restrict deletions on this record since Stash will not read it.

Maybe change the interface to support `_compactionControlSource.getOldestStashTimeForPlacement(record.getKey().getTable().getAvailability().getPlacement())`

billkalter commented on sor/src/main/java/com/bazaarvoice/emodb/sor/core/DistributedCompactor.java:
```diff
@@ -24,12 +25,12 @@ public DistributedCompactor(Counter archiveDeltaSizeInMemory, boolean keepDeltaH
     }
 
-    public Expanded expand(Record record, long fullConsistencyTimestamp, long compactionConsistencyTimeStamp, MutableIntrinsics intrinsics,
+    public Expanded expand(Record record, long fullConsistencyTimestamp, long compactionConsistencyTimeStamp, long stashRunTimeStamp, MutableIntrinsics intrinsics,
```
Consistency:  Here the timestamp is called "stashRunTimeStamp" but elsewhere it is also sometimes called "deleteDeltasTimestamp".  Recommend settling on one consistent name, maybe a middle ground between the two of something like "compactionControlTimestamp".

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/ScanUploader.java:
```diff
@@ -47,18 +47,16 @@
     private final StashStateListener _stashStateListener;
-    private final DefaultCompactionControlManager _defaultCompactionControlManager;
-    private final DataCenters _dataCenters;
+    private final List<CompactionControlSource> _compactionControlSourceList;
```
Would it make sense to pass in an implementation of `CompactionControlSource` which itself delegates to all other `CompactionControlSources`?  Here's my thinking:

Every time the list is iterated there's a chance that one of the calls will fail.  There's all sorts of failure and retry logic that goes into that, and that will have to be repeated on every use in this class.  Seems more natural to centralize that in an implementation which just mimics what you've already done here:  Iterate of the list of delegates, make the call on each, and respond to errors appropriately.

billkalter commented on sor/src/main/java/com/bazaarvoice/emodb/sor/core/DistributedCompactor.java:
```diff
@@ -158,6 +159,11 @@ protected Expanded doExpand(Record record, long fullConsistencyTimestamp, long c
             resolver = new DefaultResolver(intrinsics, compaction);
         }
+
+        // consider deleteDeltasTimestamp and not include those Ids.
```
I'd considered recommending an alternative approach with checking for `deleteDeltasTimestamp` at insertion point into `keysToDelete` rather than adding to and then removing from the list.  However, I think what you've done here is cleaner and gives a better separation of the two concerns.  My only comment is to better comment what is going on here, since to me it's non-obvious why the values inserted into `keysToDelete` are suddenly no good.

billkalter commented on sor-api/src/main/java/com/bazaarvoice/emodb/sor/api/CompactionControlSource.java:
```diff
@@ -0,0 +1,20 @@
+
+    Map<String, StashRunTimeInfo> getStashTimes();
+
+    long getOldStashTime();
```
Open question: This seems like more of a convenience method than providing any core functionality.  Would it make more sense to remove this and have the callers use `getStashTimes()` directly and manipulate however they need?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/CompactionControlMonitor.java:
```diff
@@ -0,0 +1,69 @@
+    private void deleteExpiredStashTimes() {
+        try {
+            _log.debug("Checking for expired stash times at {}", CURRENT_TIME);
+            Set<String> expiredIds = Maps.filterValues(_compactionControlSource.getAllStashTimes(), value -> value.getExpiredTimestamp() > CURRENT_TIME).keySet();
```
Can `CURRENT_TIME` be static?   Wouldn't this need to use the actual current time as returned by `System.getCurrentTimeMillis()` or, even better, an injected `Clock.millis()`?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/CompactionControlMonitor.java:
```diff
@@ -0,0 +1,69 @@
+            Set<String> expiredIds = Maps.filterValues(_compactionControlSource.getAllStashTimes(), value -> value.getExpiredTimestamp() > CURRENT_TIME).keySet();
+            for (String id : expiredIds) {
+                // If we are deleting the entries here, then there could be a problem which we may want to know. So setting it as an ERROR.
+                _log.error("Deleting the stash time entry for id: {}", id);
```
`error()` seems a bit harsh, it seems like more of a `warn()` situation to me.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/CompactionControlMonitor.java:
```diff
@@ -0,0 +1,69 @@
+            for (String id : expiredIds) {
+                // If we are deleting the entries here, then there could be a problem which we may want to know. So setting it as an ERROR.
+                _log.error("Deleting the stash time entry for id: {}", id);
+                _compactionControlSource.deleteStashTime(id, _dataCenters.getSelf().getName());
```
Is this right?  Shouldn't the data center parameter be the return from `StashRunTimeInfo#getDataCenter()` and not the local data center?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/DelegateCompactionControlSource.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public StashRunTimeInfo getStashTime(String id, String dataCenter) {
+        try {
+            for (CompactionControlSource compactionControlSource : _compactionControlSourceListProvider.get()) {
```
Does this need to iterate over all delegates?  Shouldn't it be sufficient to only query the local compaction control source?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/DelegateCompactionControlSource.java:
```diff
@@ -0,0 +1,95 @@
+    public Map<String, StashRunTimeInfo> getAllStashTimes() {
+        Map<String, StashRunTimeInfo> stashTimes = Maps.newLinkedHashMap();
+        try {
+            for (CompactionControlSource compactionControlSource : _compactionControlSourceListProvider.get()) {
```
Does this need to iterate over all delegates?  Shouldn't it be sufficient to only query the local compaction control source?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/compactioncontrol/DelegateCompactionControlSource.java:
```diff
@@ -0,0 +1,95 @@
+    public Map<String, StashRunTimeInfo> getStashTimesForPlacement(String placement) {
+        Map<String, StashRunTimeInfo> stashTimes = Maps.newLinkedHashMap();
+        try {
+            for (CompactionControlSource compactionControlSource : _compactionControlSourceListProvider.get()) {
```
Does this need to iterate over all delegates?  Shouldn't it be sufficient to only query the local compaction control source?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/resources/compactioncontrol/CompactionControlResource1.java:
```diff
@@ -0,0 +1,77 @@
+        checkArgument(!Strings.isNullOrEmpty(dataCenter), "datacenter is required");
+
+        StashRunTimeInfo stashTimeInfo = _compactionControlSource.getStashTime(id, dataCenter);
+        return (stashTimeInfo != null) ? stashTimeInfo : new StashRunTimeInfo();
```
Wouldn't it make more sense to return 404 if the stash time isn't found rather than a generic empty instance?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/ScanUploader.java:
```diff
@@ -95,22 +106,39 @@ private ScanPlan createPlan(String scanId, ScanOptions options) {
             _scanWorkflow.scanStatusUpdated(scanId);
 
-            // Send notification that the scan has started
+            // Send notification that the scan has startedcd 
```
Typo?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/rangescan/LocalRangeScanUploader.java:
```diff
@@ -243,7 +252,16 @@ public void run() {
+                cutoffTime = new DateTime(stashTimeInfoMap.entrySet().stream()
+                        .min((entry1, entry2) -> entry1.getValue().getTimestamp() > entry2.getValue().getTimestamp() ? 1 : -1)
+                        .get().getValue().getTimestamp());
+            }
```
I don't think this is the right way to check for the cutoff time.  I think it make more sense to store it with the stash state and pass it as a parameter here.  The risk is that if multiple stashes are running concurrently this will use the minimum start time of both stashes, which is incorrect if they didn't start at the same time.  This is rare but not impossible, so better to rely on the known start time for this stash rather than the control source artifact.

sujithvaddi commented:
Addressed the feedback.

sujithvaddi commented:
Jenkins retest this please

sujithvaddi commented:
Jenkins retest this please

billkalter commented:
@sujithvaddi One more thing.  In [LocalRangeScanUploader.java](https://github.com/bazaarvoice/emodb/blob/master/web/src/main/java/com/bazaarvoice/emodb/web/scanner/rangescan/LocalRangeScanUploader.java#L446) is this line:

```
Map<String, Object> content = _dataTools.toContent(result, ReadConsistency.STRONG, compactionEnabled);
```

Currently stash has compaction disabled by default but it can be enabled manually.  We need to remove this feature and change `compactionEnabled` above to always be `false`.  Since the resolved record will ignore deltas written after the stash started the `Record` returned by `result.getRecord()` above may not be the most current version of the document and cannot be used for compaction without risking data loss.

sujithvaddi commented:
@billkalter addressed the feedback, please take a look when you get a chance.

sujithvaddi commented:
Jenkins retest this please

sujithvaddi commented:
Jenkins retest this please

billkalter commented on sor/src/test/java/com/bazaarvoice/emodb/sor/core/CompactionControlTest.java:
```diff
@@ -23,104 +24,178 @@
+    }
+
+    @Test
+    public void deltasAfterCompactionControlTimestampShouldOnlyBeDeletedAndDeltasBeforeCompactionControlTimestampShouldNotBeDeleted() {
```
Now _that_ is a descriptive unit test name.

billkalter commented on sor/src/test/java/com/bazaarvoice/emodb/sor/core/CompactionControlTest.java:
```diff
@@ -23,104 +24,178 @@
+        // Verify that there are all 3 keys for deletion.
+        // In the DistributedCompactor, at first 3 keys will be selected for deletion, and later no keys get filtered out from the deletionList based on the compactionControlTimestamp.
+        assertTrue(expanded.getPendingCompaction() != null);
+        assertTrue(expanded.getPendingCompaction().getKeysToDelete().size() == 3, "All 3 deltas should up for deletion.");
```
Minor:  Checking the size of `getKeysToDelete()` is redundant since the next line does an exact content comparison.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/ScanOptions.java:
```diff
@@ -54,8 +52,7 @@ private ScanOptions(@JsonProperty ("placements") List<String> placements,
                         @JsonProperty ("rangeScanSplitSize") Integer rangeScanSplitSize,
-                        @JsonProperty ("maxRangeScanTime") Long maxRangeScanTime,
-                        @JsonProperty ("compactionEnabled") Boolean compactionEnabled) {
+                        @JsonProperty ("maxRangeScanTime") Long maxRangeScanTime) {
```
Since we're removing a field can you add the following annotation to the class:

`@JsonIgnoreProperties(ignoreUnknown = true)`

This way when it tries to read existing stash records from the Emo table it won't choke on old records where this attribute is present.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/scanstatus/ScanStatus.java:
```diff
@@ -100,6 +104,10 @@ public void setCompleteTime(Date completeTime) {
         _completeTime = completeTime;
     }
 
+    public Date getCompactionControlTime() {
```
Annotate this method with `@JsonIgnore`.  This value is computed from another value and doesn't need to go in JSON.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/rangescan/LocalRangeScanUploader.java:
```diff
@@ -186,7 +186,7 @@ public void stop()
     @Override
     public RangeScanUploaderResult scanAndUpload(
-            final int taskId, ScanOptions options, final String placement, ScanRange scanRange, final TableSet tableSet, Date stashStartTime)
+            final int taskId, ScanOptions options, final String placement, ScanRange scanRange, final TableSet tableSet, Date compactionControlTime)
```
Minor:  Can you also change the parameter name to `compactionControlTime` in the `RangeScanUploader` interface?

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/ScanUploader.java:
```diff
@@ -110,13 +109,14 @@ private void startScanUpload(String scanId, ScanStatus status) {
+            long expireTime = compactionControlTime + Duration.ofHours(10).toMillis();
             // Update the scan start time in Zookeeper in all data centers.
-            _compactionControlSource.updateStashTime(scanId, stashStartTime, Lists.newArrayList(status.getOptions().getPlacements()), expireTime, _dataCenters.getSelf().getName());
+            _compactionControlSource.updateStashTime(scanId, compactionControlTime, Lists.newArrayList(status.getOptions().getPlacements()), expireTime, _dataCenters.getSelf().getName());
```
This addresses my previous comment.  However, without a minute wait here or before the first scan this _technically_ isn't a snapshot scan.  Any record which is updated between `stashStartTime` and `compactionControlTime` one minute later will non-deterministically appear in stash in either one of the two possible resolved values.  This may fall into the "best effort snapshot" catch-all, where we explicitly declare that updates right around the time that stash starts may or may not be included in stash, but if that's the case we should update the docs making that built-in fuzziness known.

sujithvaddi commented:
@billkalter addressed the feedback. thanks!

sujithvaddi commented:
Jenkins retest this please

billkalter commented on sor/src/main/java/com/bazaarvoice/emodb/sor/core/DefaultDataStore.java:
```diff
@@ -334,12 +339,18 @@ private Resolved resolve(Record record, ReadConsistency consistency) {
+        // we are also filtering out any expired timestamps. (CompactionControlMonitor should do this for us, but for now it's running every hour. So, just to fill that gap, we are filtering here.)
+        // If no timestamps are found, then taking minimum value because we want all the deltas after the compactionControlTimestamp to be deleted as per the compaction rules as usual.
+        long compactionControlTimestamp = stashTimeInfoMap.isEmpty() ?
+                Long.MIN_VALUE : stashTimeInfoMap.values().stream().filter(s -> s.getExpiredTimestamp() > System.currentTimeMillis()).map(StashRunTimeInfo::getTimestamp).min(Long::compareTo).get();
```
There's a blind call to `.get()` at the end of this stream even though there may be no compaction control timestamps which aren't expired.  Maybe replace with `.orElse(Long.MIN_VALUE)`.

billkalter commented on web/src/main/java/com/bazaarvoice/emodb/web/scanner/ScanUploader.java:
```diff
@@ -96,33 +115,60 @@ private ScanPlan createPlan(String scanId, ScanOptions options) {
+        {
+            try {
+                // We would like to wait 1 minute here to continue to scan to make sure scan don't miss any deltas that are written before the compaction control time.
+                Thread.sleep(_compactionControlBufferTimeInMillis);
```
This sleep is necessary, but doing it here seems dangerous.  If there is a failure, such as process death while sleeping, there is nothing to ensure the scan starts.  Maybe it makes more sense to sleep after the first message is pulled from the queue.  That way if there is a failure the queued message will eventually be redelivered.
